aggregator.py

```diff
@@ -0,0 +1,183 @@
+  best to send along to the BeliefManager. """
+
+  def __init__(self, mav_queue, radio_queue, belief_box):
+    """ Args:
+      mav_queue: The queue for receiving autopilot data.
+      radio_queue: The queue for receiving radio data.
+      belief_box: The mailbox for sending data to the BeliefManager process. """
+    self.__mav_queue = mav_queue
+    self.__radio_queue = radio_queue
+    self.__belief_box = belief_box
+
+    # A somewhat arbitrary time that we define as denoting cycle zero.
+    self.__cycle_zero = time.time()
+    self.__cycle = 1
+    # These are dictionaries that store all data based on which cycle it comes
+    # from.
+    self.__future_radio = {}
+    self.__future_mav = {}
+
+  def __cycle_from_time(self, time):
+    """ Gets the cycle number we are on based on where cycle_zero denotes the
+    start of the first cycle.
+    Args:
+      time: The time we are converting to a cycle.
+    Returns:
+      The cycle corresponding to this time. """
+    cycle = (time - self.__cycle_zero) / BeliefManager.CYCLE_LENGTH
+    if cycle - int(cycle) > 0.0001:
+      # If it doesn't go in perfectly, we should add one, because it's
+      # technically in the next cycle.
+      cycle += 1
+
+    return int(cycle)
+
+  def iterate(self):
+    """ Runs a single iteration of the process. This should be synched with the
+    frequency at which BeliefManager is run. """
+    logger.debug("Running iteration %d." % (self.__cycle))
+
+    # Don't run this cycle until BeliefManager reads the data for the last
+    # cycle.
+    self.__belief_box.wait_for_read()
+
+    # Get and buffer radio data.
+    while not self.__radio_queue.empty():
+      radio_message = self.__radio_queue.get_nowait()
+
+      # Check if message is outdated.
+      message_cycle = self.__cycle_from_time(radio_message.time_created)
+      if message_cycle < self.__cycle:
+        # This really shouldn't happen at all, unless, for some reason, our
+        # queue has rediculous latency.
+        logger.warning("Dropping radio packet from previous cycle.")
+        continue
+
+      # Classify the message based on its timestamp.
+      if message_cycle > self.__cycle:
+        # This belongs in the future, so we'll save it for later.
+        logger.debug("Saving radio packet for cycle %d." % (message_cycle))
+        _make_or_add(self.__future_radio, message_cycle, radio_message)
+        continue
+
+      # It belongs in the current cycle, add it there.
+      _make_or_add(self.__future_radio, self.__cycle, radio_message)
+
+    # Put all data for this cycle into the proper format.
+    radio_buffer = []
+    cycle_data = self.__future_radio.pop(self.__cycle, [])
+    for saved_data in cycle_data:
+      radio_tuple = (saved_data.lob, saved_data.strength)
+      radio_buffer.append(radio_tuple)
+
+    if not radio_buffer:
+      logger.warning("Got no radio data for this cycle.")
+
+    # Get and process drone position data.
+    while not self.__mav_queue.empty():
```

Comment:
So, since it's relatively simple, I've decided to just put a hard limit on how many messages it can read in one iteration.